File: java/src/main/java/io/dure/tree/PostorderTraversal.java
package io.dure.tree;

import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

public class PostorderTraversal {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> path = new ArrayList<>();
        Deque<TreeNode> stack = new LinkedList<>();
        TreeNode curr = root;
        TreeNode prev = null;
        while (curr != null || !stack.isEmpty()) {
            while (curr != null) {
                stack.push(curr);
                curr = curr.left;
            }
            curr = stack.peek();
            if (curr.right != null && curr.right != prev) {
                curr = curr.right;
            } else {
                stack.pop();
                path.add(curr.val);
                prev = curr;
                curr = null;
            }
        }
        return path;
    }
}

File: go/tree/postorderTraversal.go
package tree

func postorderTraversal(root *TreeNode) []int {
	var path []int
	var stack []*TreeNode
	curr := root
	var prev *TreeNode
	for curr != nil || len(stack) != 0 {
		for curr != nil {
			stack = append(stack, curr)
			curr = curr.Left
		}
		curr = stack[len(stack)-1]
		if curr.Right != nil && curr.Right != prev {
			curr = curr.Right
		} else {
			stack = stack[:len(stack)-1]
			path = append(path, curr.Val)
			prev = curr
			curr = nil
		}
	}
	return path
}
